src/hotspot/cpu/x86/vm_version_x86.cpp

```diff
@@ -3215,6 +3215,50 @@ intx VM_Version::allocate_prefetch_distance(bool use_watermark_prefetch) {
   }
 }
 
+// Max vector size in bytes. 0 if not supported.
+int VM_Version::max_vector_size(BasicType bt) {
+  assert(is_java_primitive(bt), "only primitive type vectors");
+  if (UseSSE < 2) return 0;
+  // SSE2 supports 128bit vectors for all types.
+  // AVX2 supports 256bit vectors for all types.
+  // AVX2/EVEX supports 512bit vectors for all types.
+  int size = (UseAVX > 1) ? (1 << UseAVX) * 8 : 16;
+  // AVX1 supports 256bit vectors only for FLOAT and DOUBLE.
+  if (UseAVX > 0 && (bt == T_FLOAT || bt == T_DOUBLE))
+    size = (UseAVX > 2) ? 64 : 32;
+  if (UseAVX > 2 && (bt == T_BYTE || bt == T_SHORT || bt == T_CHAR))
+    size = (VM_Version::supports_avx512bw()) ? 64 : 32;
+  // Use flag to limit vector size.
+  size = MIN2(size,(int)MaxVectorSize);
+  // Minimum 2 values in vector (or 4 for bytes).
+  switch (bt) {
+  case T_DOUBLE:
+  case T_LONG:
+    if (size < 16) return 0;
+    break;
+  case T_FLOAT:
+  case T_INT:
+    if (size < 8) return 0;
+    break;
+  case T_BOOLEAN:
+    if (size < 4) return 0;
+    break;
+  case T_CHAR:
+    if (size < 4) return 0;
+    break;
+  case T_BYTE:
+    if (size < 4) return 0;
+    break;
+  case T_SHORT:
+    if (size < 4) return 0;
+    break;
+  default:
+    assert(false, "Unexpected basic type");
+  }
```

Comment:
If the return value is 0, what is the actual number of the multifields in payload? It seems not reasonable for `0`. WDYT?

Reply:
For x86 this leaf level routine is directly imported from Matcher.  _Matcher::max_vector_size_ will query it to get maximum vector size supported by a target. Agree that it needs to be tuned.

BTW I also see a discrepancy b/w lane count and max vector size for MaxSpecies in existing implementation.

```
LPROMPT>cat test_max.java 

import jdk.incubator.vector.*;

public class test_max {
   public static void main(String [] args) {
      System.out.println(ByteVector.SPECIES_MAX);
      System.out.println(ByteVector.broadcast(ByteVector.SPECIES_MAX, 1));
   }
} 
LPROMPT>java -XX:MaxVectorSize=0 --add-modules=jdk.incubator.vector -XX:+PrintFlagsFinal -cp . test_max.java  | grep -e "MaxVectorSize"  -e "Species" -e "\["
OpenJDK 64-Bit Server VM warning: MaxVectorSize must be at least 4 on this platform
[Global flags]
     intx MaxVectorSize                            = 4                                      {C2 product} {command line}
WARNING: Using incubator modules: jdk.incubator.vector
warning: using incubating module(s): jdk.incubator.vector
1 warning
Species[byte, 8, S_Max_BIT]
[1, 1, 1, 1, 1, 1, 1, 1]
LPROMPT>
```

Reply:
Yes, it seems werid. I think it is because the min operation with 64-bit shape in java code. Please see my following comments.